### PR TITLE
Handle form choice duplicates

### LINKED ISSUE #28878: BO - Shipping preferences - ID of the carrier is displayed in the  Default carrier list after fresh install
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

This issue id detected by nightly: https://nightly.prestashop.com/report/4307
ping @nesrineabdmouleh @boubkerbribri 
Related to this PR: https://github.com/PrestaShop/PrestaShop/pull/28718
In the BO > Shipping > Preferences page > Carrier options, the id of the carrier is always displayed.
image

### Expected behavior

The id should be displayed only when we have two carriers with the same name
image

### Steps to reproduce

1. Fresh install develop
2. Go to BO > Shipping > Preferences page > Carrier options
3. Check the list > the id of the carrier is displayed although we don't have two carriers with the same name

image

### PrestaShop version(s) where the bug happened

develop

### PHP version(s) where the bug happened

8.1

### If your bug is related to a module, specify its name and its version

_No response_

### LINKED ISSUE #36586: PPV2 - If 2 Features have the same name, only one will be displayed
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

If a Feature is duplicated, in the Product Detail page, only one feature will be displayed.
Same goes for a Feature value.

### Expected behavior

Both features/features values should be displayed.

A possible workaround would be:
https://github.com/PrestaShop/PrestaShop/blob/11ae6d7ce71888de4b911ebe745ce5d423d57004/src/Adapter/Form/ChoiceProvider/FeatureValuesChoiceProvider.php#L85

Line: 85
```
        foreach ($featureValues as $feature) {
            $this->cacheFeatureValueChoices[$cacheKey][$feature['id_feature_value']." - ".$feature['value']] = (int) $feature['id_feature_value'];
        }

```

### Steps to reproduce

1. Go to BO > Catalog > Attributes and Features > Features Tab
2. Add 2 **features** with the same name
3. Go to BO > Product page > Edit any product > Go to Details tab
4. Choose a Feature : See only one feature is displayed :x:
5. Go to BO > Catalog > Attributes and Features > Features Tab
6. Add 2 **features values** with the same name
7. Go to BO > Product page > Edit any product > Go to Details tab
8. Choose the feature where you have 2 identical feature values :
9. Choose a value : See only one feature value is displayed :x:

### PrestaShop version(s) where the bug happened

8.1.7, 9.0.0

### PHP version(s) where the bug happened

8.1

### If your bug is related to a module, specify its name and its version

_No response_

### Your company or customer's name goes here (if applicable).

_No response_

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | Introduces a class that formats choices for symfony forms and automatically marks duplicate options with an ID, so we avoid having choices that just disappear. Initally implemented on Manufacturer, Supplier, Feature value and Carrier choices, but easy to do on the rest.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | Create two manufacturers with the same name, go to product form, see that you can see both.
| UI Tests          | https://github.com/florine2623/testing_pr/actions/runs/11288952311 ✅ 
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/36586
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

**Comment by Hlavtox** [comment]
Ping @matks @jolelievre @kpodemski @FabienPapet to check the concept, before I move on. Is it OK?

**Comment by Hlavtox** [comment]
@PrestaShop/committers Hi, can someone review please? :-)

File: src/PrestaShopBundle/Resources/config/services/adapter/form.yml
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/adapter/form.yml]:
```suggestion
    autowire: true
```

File: src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml]:
```suggestion
    autowire: true
```

File: src/Adapter/Carrier/CarrierDataProvider.php
Comment (FabienPapet) [inline src/Adapter/Carrier/CarrierDataProvider.php]:
```suggestion
    private FormChoiceFormatter $formChoiceFormatter;
```

**Comment by FabienPapet** [comment]
I'd prefer having a static class instead of using DI in this specific case but it is not blocking 😀

**Comment by Hlavtox** [comment]
@FabienPapet Changed to static class and added a test. ;-)

**Review by FabienPapet — APPROVED** [review]

**Comment by FabienPapet** [comment]
I love your test data

File: tests/Unit/Core/Util/FormChoiceFormatterTest.php
Comment (FabienPapet) [inline tests/Unit/Core/Util/FormChoiceFormatterTest.php]: Copy pasted ;) not blocking

Comment (Hlavtox) [reply tests/Unit/Core/Util/FormChoiceFormatterTest.php]: Ha, good catch, I will fix it. Also copied `clean absolute url _token`.

**Review by FabienPapet — DISMISSED** [review]

**Comment by Hlavtox** [comment]
Ping @jolelievre @matks @boherm

**Review by ShaiMagal — DISMISSED** [review]

**Review by boherm — DISMISSED** [review]

**Comment by florine2623** [comment]
@Hlavtox , UI tests are red, could you check that ? 
https://github.com/florine2623/testing_pr/actions/runs/11105865072

@PrestaShop/qa-automation , what do you think about this ? 

Thanks to all of you !

**Comment by Progi1984** [comment]
@florine2623 The nightly is :green_circle: so it's related to the PR.

image

File: src/Core/Util/FormChoiceFormatter.php
Comment (jolelievre) [inline src/Core/Util/FormChoiceFormatter.php]:
```suggestion
namespace PrestaShop\PrestaShop\Core\Form;
```

Comment (jolelievre) [reply src/Core/Util/FormChoiceFormatter.php]: Form namespace seems more accurate for this tool

File: src/Core/Util/FormChoiceFormatter.php
Comment (jolelievre) [inline src/Core/Util/FormChoiceFormatter.php]:
```suggestion
        $alreadyProcessedNames = [];
```

File: src/Core/Util/FormChoiceFormatter.php
Comment (jolelievre) [inline src/Core/Util/FormChoiceFormatter.php]: Not sure this is needed or if it brings any improvement on performances 🤔 You can simply loop through the array without modifying it If so you can also simplify the for loop:
```php
foreach ($rawChoices as $rawChoice) {
```

File: tests/Unit/Core/Util/FormChoiceFormatterTest.php
Comment (jolelievre) [inline tests/Unit/Core/Util/FormChoiceFormatterTest.php]:
```suggestion
namespace Tests\Unit\Core\Form;
```

File: tests/Unit/Core/Util/FormChoiceFormatterTest.php
Comment (jolelievre) [inline tests/Unit/Core/Util/FormChoiceFormatterTest.php]:
```suggestion
    public function testFormatFormChoices(array $rawOptions, string $idKey, string $nameKey, array $expectedFormattedChoices): void
```

File: tests/Unit/Core/Util/FormChoiceFormatterTest.php
Comment (jolelievre) [inline tests/Unit/Core/Util/FormChoiceFormatterTest.php]:
```suggestion
        yield 'manufacturer list with duplicates' => [
```

File: tests/Unit/Core/Util/FormChoiceFormatterTest.php
Comment (jolelievre) [inline tests/Unit/Core/Util/FormChoiceFormatterTest.php]: You should also add an use case where there is no duplicates, to validate that in this case the names are not modified at all (which is 90% of the time probably)

File: src/Core/Util/FormChoiceFormatter.php
Comment (jolelievre) [inline src/Core/Util/FormChoiceFormatter.php]: Maybe, the sorting should be optional? With a `$sortByNames = true` by default, but in some cases the order can be made based on other criteria of importance and we shouldn't change it

If you add this option then a third use case with this option at `false` would be nice in the unit tests

**Review by jolelievre — COMMENTED** [review]
Thanks @Hlavtox 

That's a great idea! Just a few suggestions, but I love the idea!

File: src/Core/Util/FormChoiceFormatter.php
Comment (jolelievre) [reply src/Core/Util/FormChoiceFormatter.php]: Oh actually what I thought as a suggestion for a potential use case is already required, the Carrier list has a specific order by position So in this case you need to disable the automatic ksort It is likely what makes the UI tests fail in this case

Comment (jolelievre) [reply src/Core/Util/FormChoiceFormatter.php]: The other failing test maybe be related to `ProfileByIdChoiceProvider` that also forces an order by ID internally, in this case I don't know if we should keep the original order or adapt the UI tests (but for carrier the order by position makes sense and should be prioritized)

File: src/Core/Util/FormChoiceFormatter.php
Comment (Hlavtox) [reply src/Core/Util/FormChoiceFormatter.php]: My reasoning was that I gradually cut from one array and add it, so we don't increase memory usage by looping it. :-)

**Comment by Hlavtox** [comment]
@jolelievre Addressed all feedback! :-)

- Added option to not use sorting.
- For this reason, I changed the way I replace previous values in the array, I now change the key keeping the position.
- Changed namespaces.
- Changed forgotten copy-pastes in tests.
- Added new tests, without duplicates and with duplicates and no sorting.

Let's see what the tests say! https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11133538054

**Review by FabienPapet — DISMISSED** [review]

**Comment by PululuK** [comment]
Bien vu @Hlavtox 👍🏿

**Comment by Hlavtox** [comment]
I removed some usages to not have to deal with tests now, they need to be adapter. It can be done in a second step where we implement this everywhere.

New test run - https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11141582905

**Review by FabienPapet — DISMISSED** [review]

File: tests/Unit/Core/Form/FormChoiceFormatterTest.php
Comment (jolelievre) [inline tests/Unit/Core/Form/FormChoiceFormatterTest.php:151]: Nide one! 👍

**Review by jolelievre — DISMISSED** [review]
Thanks @Hlavtox

File: src/Core/Form/FormChoiceFormatter.php
Comment (PululuK) [inline src/Core/Form/FormChoiceFormatter.php:51]: I think the method name is a bit redundant with the class name. `FormChoiceFormatter::format` would be sufficient

**Review by PululuK — DISMISSED** [review]

**Comment by florine2623** [comment]
Hello @Hlavtox and @PrestaShop/qa-automation 

I have relaunched the auto tests : https://github.com/florine2623/testing_pr/actions/runs/11288952311
They are still red. What should be done from there ?

Thanks!

**Comment by florine2623** [comment]
@Hlavtox , seen with @nesrineabdmouleh , the problem of the auto test comes from your PR.

Could you check it please ? :)
Thanks!

**Comment by nesrineabdmouleh** [comment]
Hello @florine2623,

I fixed UI tests, the country ID is changed in Payment>References page, please restart the failed test :)

Thanks!

**Comment by florine2623** [comment]
Hello @Hlavtox ,

https://github.com/florine2623/testing_pr/actions/runs/11288952311 tests are green, thanks @nesrineabdmouleh !

It is QA ✅

**Review by Progi1984 — DISMISSED** [review]
Ok for ui tests

**Comment by nicosomb** [comment]
PHP CS fixer is not happy.

**Review by kpodemski — DISMISSED** [review]
Hello

We recently introduced the 9.0.x branch, which will be the base for PrestaShop 9.0. If you want to see your change in this version, please rebase your PR onto 9.0.x. The `develop` will be used for PrestaShop 9.1, so you need to decide what to do.

Please rebase onto `9.0.x` or let me know your decision if you don't need to have it there so I can unblock the pull request.

**Comment by Hlavtox** [comment]
@kpodemski Ready, code style fixed. Can we merge?

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
@Hlavtox CI, second approval and I think we can merge because it had QA ✅ before the rebase

**Comment by Hlavtox** [comment]
@kpodemski I did not rebase, just changed the branch here in Github, and it worked :-)

**Review by jolelievre — APPROVED** [review]
Thanks @Hlavtox

**Pull request merged**
